File: solutions/2197-decode-the-slanted-ciphertext/decode-the-slanted-ciphertext.java
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if(encodedText.length()<=1) return encodedText;
        int col=encodedText.length()/rows;
        char[][] mat=new char[rows][col];
        int eIndex=0;
        for(int i=0;i<rows;i++){

            for(int j=0;j<col;j++){
                mat[i][j]=encodedText.charAt(eIndex++);

            }
        }
        // display(mat);
        // String res="";
        StringBuilder res = new StringBuilder();

        for(int j=0;j<col;j++){
            int t=j;

            for(int i=0;i<rows;i++){
                if(i==rows || t==col) break;
                res.append(mat[i][t++]);
            }
        }
        // int x =0;
        // for(int k=res.length()-1;k>=0;k--){
        //     if(res.charAt(k)> 96 && res.charAt(k)<123) {
        //         x=k;
        //         break;
        //     }

        // }

          return res.toString().replaceAll("\\s+$", "");
    }
    public void display(char[][] arr){
        
        for(int i=0;i<arr.length;i++){

            for(int j=0;j<arr[0].length;j++){
                System.out.print(arr[i][j]+" ");
            }
            System.out.println();

        }
    }
}

File: solutions/2197-decode-the-slanted-ciphertext/decode-the-slanted-ciphertext.cpp
class Solution {
public:
    string decodeCiphertext(string encodedText, int rows) 
    {
        if(encodedText.empty())
        {
            return "";
        }
        
        string ans = "";
        int cols = encodedText.size()/rows;
        int i = 0;
        int j = 0;
        int n = rows;
        int m = cols;

        while(j<m)
        {
            int l = i;
            int r = j;
            while(l<n && r<m)
            {
                int id = (m*l) + r;
                ans+=encodedText[id];
                l++;
                r++;
            }
            j++;
        }

        int id = ans.length()-1;
        while(ans[id]==' ' && id>=0)
        {
            id--;
        }
        id++;

        return ans.substr(0,id);
    }
};